#9975 Add info to setup the dev env for TDD.

adiroiban opened the pull request:
Fixes #9975 

These are some notes that I took while trying to get my first patch 

I think that it helps to mention rustup, so that people will not have to mess with `rust` package from their OS... the Ubuntu one is not great.

Also, doing TDD via nox, is very slow. So I added a few notes about running in edit mode.

adiroiban commented:
There is something with docs build on Ubuntu 23.10

So the nox env is not always working.

But for this PR, the sphinx part is ok.

```
$ nox -e docs

Document: x509/ocsp
-------------------
1 items passed all tests:
  30 tests in default
30 tests in 1 items.
30 passed and 0 failed.
Test passed.

Document: hazmat/primitives/aead
--------------------------------

Traceback (most recent call last):
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/cmd/build.py", line 298, in build_main
    app.build(args.force_all, args.filenames)
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/application.py", line 355, in build
    self.builder.build_update()
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/builders/__init__.py", line 363, in build
    self.write(docnames, list(updated_docnames), method)
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/ext/doctest.py", line 360, in write
    self.test_doc(docname, doctree)
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/ext/doctest.py", line 467, in test_doc
    self.test_group(group)
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/ext/doctest.py", line 552, in test_group
    self.test_runner.run(test, out=self._warn_out, clear_globs=False)
  File "/usr/lib/python3.11/doctest.py", line 1498, in run
    return self.__run(test, compileflags, out)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/doctest.py", line 1403, in __run
    self.report_unexpected_exception(out, test, example,
  File "/usr/lib/python3.11/doctest.py", line 1271, in report_unexpected_exception
    out(self._failure_header(test, example) +
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/ext/doctest.py", line 319, in _warn_out
    logger.warning(text)
  File "/usr/lib/python3.11/logging/__init__.py", line 1855, in warning
    self.log(WARNING, msg, *args, **kwargs)
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/util/logging.py", line 127, in log
    super().log(level, msg, *args, **kwargs)
  File "/usr/lib/python3.11/logging/__init__.py", line 1887, in log
    self.logger.log(level, msg, *args, **kwargs)
  File "/usr/lib/python3.11/logging/__init__.py", line 1559, in log
    self._log(level, msg, args, **kwargs)
  File "/usr/lib/python3.11/logging/__init__.py", line 1634, in _log
    self.handle(record)
  File "/usr/lib/python3.11/logging/__init__.py", line 1644, in handle
    self.callHandlers(record)
  File "/usr/lib/python3.11/logging/__init__.py", line 1706, in callHandlers
    hdlr.handle(record)
  File "/usr/lib/python3.11/logging/__init__.py", line 974, in handle
    rv = self.filter(record)
         ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/logging/__init__.py", line 830, in filter
    result = f.filter(record)
             ^^^^^^^^^^^^^^^^
  File "/home/adi/chevah/cryptography/.nox/docs/lib/python3.11/site-packages/sphinx/util/logging.py", line 427, in filter
    raise exc
sphinx.errors.SphinxWarning: **********************************************************************
File "hazmat/primitives/aead.rst", line 187, in default
Failed example:
    aesgcmsiv = AESGCMSIV(key)
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.11/doctest.py", line 1351, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest default[5]>", line 1, in <module>
        aesgcmsiv = AESGCMSIV(key)
                    ^^^^^^^^^^^^^^
    cryptography.exceptions.UnsupportedAlgorithm: AES-GCM-SIV is not supported by this version of OpenSSL


Warning, treated as error:
**********************************************************************
File "hazmat/primitives/aead.rst", line 187, in default
Failed example:
    aesgcmsiv = AESGCMSIV(key)
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.11/doctest.py", line 1351, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest default[5]>", line 1, in <module>
        aesgcmsiv = AESGCMSIV(key)
                    ^^^^^^^^^^^^^^
    cryptography.exceptions.UnsupportedAlgorithm: AES-GCM-SIV is not supported by this version of OpenSSL

nox > Command sphinx-build -T -W -b doctest -d .nox/docs/tmp/doctrees docs docs/_build/html failed with exit code 2
nox > Session docs failed.
```

alex commented on docs/development/getting-started.rst:
```diff
@@ -17,13 +17,32 @@ installed with ``pip``.
     $ nox -e tests -p py310
 
+
+Rust dependencies on Linux
```
This is duplicative of the extensive docs we have in the installation page (which covers many platforms).

It'd be better to just link https://cryptography.io/en/latest/installation/#rust

adiroiban replied:
Yes. True. Thanks.

I now see https://cryptography.io/en/latest/installation/#rust
I missed it on the first read... as I was focued on the dev part ... install would come later :)

adiroiban replied:
The thing is that I tried `cargo` via apt-get install on Ubuntu, but it was intalled without `fmt` or `clippy` or `llvm-tools` 
so `nox -e rust` was failing.

alex commented on docs/development/getting-started.rst:
```diff
@@ -44,6 +63,27 @@ You can also specify a subset of tests to run as positional arguments:
     $ # run the whole x509 testsuite, plus the fernet tests
     $ nox -e tests -p py310 -- tests/x509/ tests/test_fernet.py
 
+For TDD you might want to speed up the test run cycle.
+After activating your virtualenv, install the dev requirements
+and then install the project in edit mode.
+
+.. code-block:: console
+
+    $ pip install -e vectors/
+    $ pip install -e .[ssh,test,test-randomorder]
+
+If you are changing the OpenSSL bindings from `src/_cffi_src` you will need
+to rebuild the rust OpenSSL bindings.
+The easiest method is to just reinstall in edit mode.
+
+.. code-block:: console
+
+    $ # Re-build the binding and other extensions.
+    $ pip install -e .
+    $ # Then re-run the tests.
+    $ pytest tests/hazmat/bindings/
+
+
```
I'm opposed to documenting this. While it works (and it's how many core devs develop), it's far too easy to screw up: notably failing to rebuild any extension module will get you confusing and inconsistent results, that's why our documentation is about `nox`.

adiroiban replied:
Maybe add a warning before this section ?

alex replied:
Unless @reaperhulk disagrees with me, I don't want to add it.

Perhaps we can suggest `nox -e tests-nocoverage` (which reduces some time), or the usage of `--no-install`, which can be used on subsequent invocations if you haven't changed any extension modules.

adiroiban replied:
No problem.

I now know how to do TDD for crypography.
I was thinking that this info is useful for others.

-------

Disabling `-n auto` can also save 6 seconds.

alex replied:
That's interesting that disable parallelism saves time, what sort of CPU do you have? I almost always run the full test suite, and it's a big time savings for me (10 cores; 2 of which are efficiency cores).

adiroiban replied:
This is only for TDD... when you run a subset of the tests

```
$ time pytest -n auto --dist=worksteal --cov=cryptography --cov=tests --durations=10 tests/hazmat/bindings/test_openssl.py

real    0m6.638s
user    0m27.316s
sys     0m3.032s

vs

$ time pytest tests/hazmat/bindings/test_openssl.py

real    0m0.587s
user    0m0.522s
sys     0m0.070s
```

adiroiban commented:
Thanks Alex for the review. I have updated the docs with latest changes with `main` and added a warning.

I understand that `nox` is the safest way... but it's slow.

And slow TDD cycles are not fun.

And if you mention that other devs are not using `nox`... then I thing that this is a sign that there is a problem with `nox`.

Now... this might not be the best solution...but is a solution to slow `nox` run.

I am happy to use any other faster method.

This is just my feedback from trying to get `cryptography` dev env working on Ubuntu.

Feel free to reject the PR.

adiroiban commented:
It looks like [linkcheck](https://github.com/pyca/cryptography/blob/main/.github/workflows/linkcheck.yml) is skipped on most PRs as it is only executed on ``conf.py`` changes 

```
$ nox -e docs-linkcheck

hazmat/primitives/cryptographic-hashes: line  247) ok        https://www.oscca.gov.cn/sca/xxgk/2010-12/17/1002389/files/302a3ada057c4a73830536d03e683110.pdf
( doing-a-release: line   21) 
Warning, treated as error:
/home/adi/chevah/cryptography/docs/doing-a-release.rst:21:broken link: https://www.openwall.com/lists/oss-security/ (HTTPSConnectionPool(host='www.openwall.com', port=443): Max retries exceeded with url: /lists/oss-security/ (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7ff0dca76e90>: Failed to establish a new connection: [Errno 113] No route to host')))
nox > Command sphinx-build -W -b linkcheck docs docs/_build/html failed with exit code 2
nox > Session docs-linkcheck failed.
```

alex commented:
Thank you for the PR. We made a bunch of improvements to the devdocs based on this. However, after discussing with @reaperhulk we both agree that we do not want to document the non-nox local workflow.
